File: packages/core/go/secretstore/backend.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// The SecretBackend read surface for the Go data plane: every provider credential read goes through this interface.

package secretstore

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"
)

// Backend resolves opaque secret values addressed by hierarchical refs. The data
// plane only reads, so external backend credentials issued to Go services can be
// scoped read-only; all writes happen in the control plane.
type Backend interface {
	Kind() string
	Get(ctx context.Context, ref string) (value []byte, found bool, err error)
}

// Kinds accepted by CARACAL_SECRET_BACKEND. The builtin backend is constructed by
// the owning service because it is bound to that service's database pool.
const (
	KindBuiltin           = "builtin"
	KindVault             = "vault"
	KindInfisical         = "infisical"
	KindAzureKeyVault     = "azurekeyvault"
	KindAWSSecretsManager = "awssecretsmanager"
	KindGCPSecretManager  = "gcpsecretmanager"
	KindCustom            = "custom"
)

// KindFromEnv resolves the configured backend kind, defaulting to builtin.
func KindFromEnv(raw string) (string, error) {
	kind := strings.ToLower(strings.TrimSpace(raw))
	if kind == "" {
		return KindBuiltin, nil
	}
	switch kind {
	case KindBuiltin, KindVault, KindInfisical, KindAzureKeyVault, KindAWSSecretsManager, KindGCPSecretManager, KindCustom:
		return kind, nil
	default:
		return "", fmt.Errorf("CARACAL_SECRET_BACKEND must be one of builtin, vault, infisical, azurekeyvault, awssecretsmanager, gcpsecretmanager, custom; got %q", kind)
	}
}

// ProviderSecretConfigRef addresses a provider's sealed credential document.
func ProviderSecretConfigRef(zoneID, providerID string) string {
	return "zones/" + zoneID + "/providers/" + providerID + "/secretConfig"
}

// AAD strings for machine-generated runtime material sealed with the builtin
// envelope in its owning table. Fixed per column family so control-plane row
// rewrites and data-plane reads agree; these must match the TypeScript constants.
const (
	AADConnectionAccessToken  = "caracal/providerConnections/accessToken"
	AADConnectionRefreshToken = "caracal/providerConnections/refreshToken"
	AADZoneSigningKey         = "caracal/secrets/zoneSigningKey"
)

type cacheEntry struct {
	value     []byte
	found     bool
	expiresAt time.Time
}

type cachedBackend struct {
	inner Backend
	ttl   time.Duration
	mu    sync.RWMutex
	byRef map[string]cacheEntry
}

// NewCached bounds external backend read traffic with a short in-memory TTL, so hot
// token-exchange paths do not turn every mint into a network round trip. Credential
// rotation propagates within one TTL.
func NewCached(inner Backend, ttl time.Duration) Backend {
	return &cachedBackend{inner: inner, ttl: ttl, byRef: make(map[string]cacheEntry)}
}

func (c *cachedBackend) Kind() string { return c.inner.Kind() }

func (c *cachedBackend) Get(ctx context.Context, ref string) ([]byte, bool, error) {
	c.mu.RLock()
	entry, ok := c.byRef[ref]
	c.mu.RUnlock()
	if ok && time.Now().Before(entry.expiresAt) {
		return entry.value, entry.found, nil
	}
	value, found, err := c.inner.Get(ctx, ref)
	if err != nil {
		return nil, false, err
	}
	c.mu.Lock()
	c.byRef[ref] = cacheEntry{value: value, found: found, expiresAt: time.Now().Add(c.ttl)}
	c.mu.Unlock()
	return value, found, nil
}

File: packages/serverCore/ts/src/secretStore/backend.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// The SecretBackend abstraction: every durable credential write and read in Caracal goes through this interface.

import { openSecretEnvelope, sealSecretEnvelope } from './envelope.js'

export const SECRET_BACKEND_KINDS = [
  'builtin',
  'vault',
  'infisical',
  'azurekeyvault',
  'awssecretsmanager',
  'gcpsecretmanager',
  'custom',
] as const

export type SecretBackendKind = (typeof SECRET_BACKEND_KINDS)[number]

// A backend stores opaque secret values addressed by hierarchical refs. The control
// plane holds the full surface; the data plane only ever reads, so external backend
// credentials issued to the STS can be scoped read-only.
export interface SecretBackend {
  readonly kind: SecretBackendKind
  put(ref: string, value: Buffer): Promise<void>
  get(ref: string): Promise<Buffer | null>
  delete(ref: string): Promise<void>
}

// Raised when the configured backend cannot serve a request; callers map this to a
// service-unavailable response instead of leaking backend specifics.
export class SecretBackendError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'SecretBackendError'
  }
}

// The crypto boundary of the Secret Store: every value is sealed into a CSS1
// envelope before it reaches a backend and opened after it comes back, with the
// ref as AAD. Backends only ever carry ciphertext, so an external manager holds
// envelopes it cannot read and the master KEK never leaves the service.
export class SealedBackend implements SecretBackend {
  constructor(private readonly inner: SecretBackend) {}

  get kind(): SecretBackendKind {
    return this.inner.kind
  }

  async put(ref: string, value: Buffer): Promise<void> {
    await this.inner.put(ref, sealSecretEnvelope(value, ref))
  }

  async get(ref: string): Promise<Buffer | null> {
    const envelope = await this.inner.get(ref)
    if (envelope === null) return null
    return openSecretEnvelope(envelope, ref)
  }

  async delete(ref: string): Promise<void> {
    await this.inner.delete(ref)
  }
}

export function secretBackendKind(): SecretBackendKind {
  const raw = (process.env.CARACAL_SECRET_BACKEND || 'builtin').trim().toLowerCase()
  if ((SECRET_BACKEND_KINDS as readonly string[]).includes(raw)) return raw as SecretBackendKind
  throw new Error(`CARACAL_SECRET_BACKEND must be one of ${SECRET_BACKEND_KINDS.join(', ')}, got '${raw}'`)
}

export function providerSecretConfigRef(zoneId: string, providerId: string): string {
  return `zones/${zoneId}/providers/${providerId}/secretConfig`
}

// AAD strings for machine-generated runtime material sealed with the builtin envelope
// in its owning table. Fixed per column family so control-plane writes and data-plane
// reads agree; the connection token strings must match the Go constants.
export const AAD_CONNECTION_ACCESS_TOKEN = 'caracal/providerConnections/accessToken'
export const AAD_CONNECTION_REFRESH_TOKEN = 'caracal/providerConnections/refreshToken'
export const AAD_NOTIFICATION_SINK_SECRET = 'caracal/notificationSinks/secret'
export const AAD_OPERATOR_PLAN_SECRETS = 'caracal/operatorPlanSecrets/values'
export const AAD_ZONE_SIGNING_KEY = 'caracal/secrets/zoneSigningKey'
